File: javascript/easy/twoSumIV.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {boolean}
 */
var findTarget = function(root, k) {
    const list = new Array();
    inOrder(root, list);
    let i = 0, j = list.length-1;
    while(i < j) {
        let sum = list[i] + list[j];
        if(sum === k) {
            return true;
        } else if(sum < k) {
            i++;
        } else {
            j--;
        }
    }
    return false;
};

var inOrder = function(root, list) {
    if(root === null) return;
    inOrder(root.left, list);
    list.push(root.val);
    inOrder(root.right, list);
}

File: ruby/easy/two_sum_iv.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Integer} k
# @return {Boolean}
def find_target(root, k)
    set = Set.new
    return target_exists?(root, k, set)
end

def target_exists?(root, k, set)
    return false unless root
    return true if set.include?(k - root.val)
    set.add(root.val)
    return target_exists?(root.left, k, set) || target_exists?(root.right, k, set)
end
